### ISSUE TITLE
Q: mnist.py model classification output is not using softmax?

### ISSUE THREAD

**Issue by mw66** [1ssb/torchkan#9]
Hi,

I just realized that the mnist.py model classification output is not using softmax?

https://github.com/1ssb/torchkan/blob/main/mnist.py

and the output values are all negative values < 0.0

I'm just wondering if this is intended? maybe worth to doc this info in the readme?

Thanks.

**Comment by 1ssb** [comment]
That's interesting, sorry I missed that. Thanks for raising the issue. I will look into it soon.

Best
Subhransu
________________________________
From: mw66 ***@***.***>
Sent: Friday, June 28, 2024 5:06:38 PM
To: 1ssb/torchkan ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [1ssb/torchkan] Q: mnist.py model classification output is not using softmax? (Issue #9)


Hi,

I just realized that the mnist.py model classification output is not using softmax?

https://github.com/1ssb/torchkan/blob/main/mnist.py

and the output values are all negative values < 0.0

I'm just wondering if this is intended? maybe worth to doc this info in the readme?

Thanks.

**Comment by 1ssb** [comment]
Okay, can you explain in detail the conditions you may be using it for. Remember that the `mnist.py` is updated and modified for every other model, so you need to update it accordingly for the kind of model you are using it on. And which "output values" do you get as negative? I might have missed something, could you be more exact?

In case you mean the softmax for the classification, note that `nn.CrossEntropyLoss()` takes care of that internally and what you maybe printing intermediately are the raw logits?

Maybe use softmax explicitly, perhaps for a different analysis where you need the probabilities, you would typically apply it directly using `torch.nn.functional.softmax` or `torch.nn.Softmax` on the logits, outside of the loss computation. For example:

`probabilities = F.softmax(output, dim=1)`

This would be necessary if you need to analyze or utilize the class probabilities directly, such as when you want to report the confidence of the model in its predictions.

**Comment by mw66** [comment]
https://github.com/1ssb/torchkan/blob/main/mnist.py#L61

I'm just printing out the `output` value after this line, and found it's not softmax (where you used argmax on it.)

**Comment by 1ssb** [comment]
Yes, so those are logits and they do not mean anything unless it's going through the labeling and finally the cross-entropy computation. 

This is not an issue of the implementation so I am going to close this issue. If you have any further concerns please mail me at: Subhransu.Bhattacharjee@anu.edu.au
